libsolidity/analysis/TypeChecker.cpp

```diff
@@ -4035,24 +4081,81 @@ void TypeChecker::endVisit(Literal const& _literal)
 						),
 						"Functions that take 2 arguments can only be used as literal suffixes for rational numbers."
 					);
-				else
+				else if (
+					dynamic_cast<IntegerType const*>(suffixFunctionType->parameterTypes().at(0)) &&
+					dynamic_cast<IntegerType const*>(suffixFunctionType->parameterTypes().at(1))
+				)
 				{
 					auto&& [mantissa, exponent] = literalRationalType->mantissaExponent();
 					// This was already validated in visit(Literal) but the error is not fatal.
 					if (!mantissa || !exponent)
 						solAssert(!m_errorReporter.errors().empty());
-					else if (
-						!mantissa->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(0)) ||
-						!exponent->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(1))
-					)
-						// TODO: Is this triggered when the argument is out of range? Test.
-						parameterTypeMessage = "The type of the literal cannot be converted to the parameters of the suffix function.";
+					// TODO: Is this triggered when the argument is out of range? Test.
+					else
+					{
+						vector<string> mantissaOrExponentErrorMessages;
+						if (!mantissa->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(0)))
+							mantissaOrExponentErrorMessages.emplace_back(
+								fmt::format(
+									"The mantissa is out of range of type {}.",
+									suffixFunctionType->parameterTypes().at(0)->toString(true)
+								)
+							);
+						if (!exponent->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(1)))
+							mantissaOrExponentErrorMessages.emplace_back(
+								fmt::format(
+									"The exponent is out of range of type {}.",
+									suffixFunctionType->parameterTypes().at(1)->toString(true)
+								)
+							);
+
+						if (!mantissaOrExponentErrorMessages.empty())
+							parameterTypeMessage =
+								"This number cannot be decomposed into a mantissa and decimal exponent "
+								"that fit the range of parameters of the suffix function. " +
+								joinHumanReadable(mantissaOrExponentErrorMessages, " ");
+					}
 				}
+				else
+					// visit(FunctionDefinition) should have spotted if any of the parameters is not IntegerType
+					solAssert(!m_errorReporter.errors().empty());
 			}
 			else if (suffixFunctionType->parameterTypes().size() == 1)
 			{
 				if (!literalType->isImplicitlyConvertibleTo(*suffixFunctionType->parameterTypes().at(0)))
-					parameterTypeMessage = "The type of the literal cannot be converted to the parameter of the suffix function.";
+				{
+					string conversionErrorMessage;
+					if (
+						dynamic_cast<IntegerType const*>(suffixFunctionType->parameterTypes().at(0)) &&
+						literalRationalType
+					)
+					{
+						conversionErrorMessage = "The number";
+						if (literalRationalType->isFractional())
+							conversionErrorMessage += " cannot be converted to";
+						else
+							conversionErrorMessage += " is out of range of";
```

Comment:
Although this fixes the problem of saying `is out of range` when the number is fractional, it adds some code duplication.
I will go back to this later and try to refactor it in a better way.

Reply:
You could define two intermediate variables (`reason` and `literalDescription`), then have two independent `if`s to fill them out and later insert them into the message template.

Reply:
Thanks for the suggestion. Done.